### PR TITLE
Add init action for metastore-grpc-proxy

### PR DESCRIPTION
**Pull request by wwlian**

### CODE REVIEW

File: metastore-grpc-proxy/metastore-grpc-proxy.sh [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    -p 5005:5005 \
+    -e HIVE_HOME=/etc/hive \
+    -v "/etc/hive/conf/hive-site.xml:/etc/hive/conf/hive-site.xml:ro" \
+    -v "/etc/hadoop/conf/core-site.xml:/etc/hadoop/conf/core-site.xml:ro" \
+    ${ADDITIONAL_DOCKER_VOLUMES} \
+    --name="${CONTAINER_NAME}" "${DOCKER_IMAGE}:${TAG}" \
+    --conf proxy.mode=${PROXY_MODE} ${COMMON_PROXY_ARG}
+}
+
+function kill_master_hms() {
```
Comment (medb) [inline metastore-grpc-proxy/metastore-grpc-proxy.sh:73]: Rename to `stop_...`

File: metastore-grpc-proxy/metastore-grpc-proxy.sh [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      return 0
+    fi
+    sleep 5
+  done
+  return 1
+}
+
+function main() {
+  if [[ "${ROLE}" == 'Master' ]]; then
+    kill_master_hms
```
Comment (medb) [inline metastore-grpc-proxy/metastore-grpc-proxy.sh:94]: Do we need to do this? I think Dataproc cluster will not have local HMS if DPMS is configured.

File: metastore-grpc-proxy/metastore-grpc-proxy.sh [hunk 3]
```diff
@@ -0,0 +1,103 @@
+#!/bin/bash
```
Comment (medb) [inline metastore-grpc-proxy/metastore-grpc-proxy.sh:1]: Please, add a test for this init action: https://github.com/GoogleCloudDataproc/initialization-actions/tree/master/integration_tests

See BigTable init action for test example: https://github.com/GoogleCloudDataproc/initialization-actions/blob/master/bigtable/test_bigtable.py

**Review by medb — CHANGES_REQUESTED** [review]

File: metastore-grpc-proxy/metastore-grpc-proxy.sh (hunk 2 above)
Comment (wwlian) [reply metastore-grpc-proxy/metastore-grpc-proxy.sh:94]: That would be the case if the Dataproc cluster were created with a [MetastoreConfig](https://cloud.google.com/dataproc/docs/reference/rest/v1/ClusterConfig#MetastoreConfig), but MetastoreConfig only works for Thrift-based DPMS services. When using this init action, the local HMS is not suppressed.

File: metastore-grpc-proxy/metastore-grpc-proxy.sh (hunk 1 above)
Comment (wwlian) [reply metastore-grpc-proxy/metastore-grpc-proxy.sh:73]: Done.

File: metastore-grpc-proxy/metastore-grpc-proxy.sh [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+function stop_master_hms() {
+  systemctl stop hive-metastore.service
+}
+
+function restart_hive_server() {
+  systemctl restart hive-server2.service
+}
+
+function wait_for_hive_server_init() {
```
Comment (cnauroth) [inline metastore-grpc-proxy/metastore-grpc-proxy.sh:81]: Is this function called from anywhere?  If not, can it be removed?

Comment (wwlian) [reply metastore-grpc-proxy/metastore-grpc-proxy.sh]: Thanks, removed.

File: metastore-grpc-proxy/metastore-grpc-proxy.sh (hunk 3 above)
Comment (wwlian) [reply metastore-grpc-proxy/metastore-grpc-proxy.sh:1]: Done

File: metastore-grpc-proxy/BUILD [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package(default_visibility = ["//visibility:public"])
+
+py_test(
```
Comment (medb) [inline metastore-grpc-proxy/BUILD:3]: You need to include test in the test suite: https://github.com/GoogleCloudDataproc/initialization-actions/blob/dc0185effc7800218da11fbce3c355ed05498815/BUILD#L3-L42

**Review by medb — CHANGES_REQUESTED** [review]

File: metastore-grpc-proxy/BUILD (hunk 5 above)
Comment (wwlian) [reply metastore-grpc-proxy/BUILD:3]: Done, thanks

File: metastore-grpc-proxy/metastore-grpc-proxy.sh [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+# This script installs a proxy that will allow a Google Cloud Dataproc cluster
+# to use a Hive Metastore that exposes a gRPC endpoint.
+
+set -euxo pipefail
+
+readonly ROLE="$(/usr/share/google/get_metadata_value attributes/dataproc-role)"
+readonly DATAPROC_MASTER=$(/usr/share/google/get_metadata_value attributes/dataproc-master)
+readonly CONTAINER_NAME=$(/usr/share/google/get_metadata_value attributes/container-name || echo "hms-proxy")
+readonly TAG=$(/usr/share/google/get_metadata_value attributes/tag || echo "v0.0.2")
```
Comment (medb) [inline metastore-grpc-proxy/metastore-grpc-proxy.sh:23]: This is too generic attribute name, rename to be more specific.

File: metastore-grpc-proxy/metastore-grpc-proxy.sh [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+readonly PROXY_MODE=$(/usr/share/google/get_metadata_value attributes/proxy-mode || echo "thrift")
+# This is what the T2G proxy connects to. Set this to the DPMS gRPC endpoint.
+readonly PROXY_URI=$(/usr/share/google/get_metadata_value attributes/proxy-uri | sed -e 's/^https:\/\///')
```
Comment (medb) [inline metastore-grpc-proxy/metastore-grpc-proxy.sh:26]: Too generic names

File: metastore-grpc-proxy/metastore-grpc-proxy.sh [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+# limitations under the License.
+
+# This script installs a proxy that will allow a Google Cloud Dataproc cluster
+# to use a Hive Metastore that exposes a gRPC endpoint.
+
+set -euxo pipefail
+
+readonly ROLE="$(/usr/share/google/get_metadata_value attributes/dataproc-role)"
+readonly DATAPROC_MASTER=$(/usr/share/google/get_metadata_value attributes/dataproc-master)
+readonly CONTAINER_NAME=$(/usr/share/google/get_metadata_value attributes/container-name || echo "hms-proxy")
```
Comment (medb) [inline metastore-grpc-proxy/metastore-grpc-proxy.sh:22]: Too generic name

File: metastore-grpc-proxy/metastore-grpc-proxy.sh [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+readonly TAG=$(/usr/share/google/get_metadata_value attributes/tag || echo "v0.0.2")
+readonly PROXY_MODE=$(/usr/share/google/get_metadata_value attributes/proxy-mode || echo "thrift")
+# This is what the T2G proxy connects to. Set this to the DPMS gRPC endpoint.
+readonly PROXY_URI=$(/usr/share/google/get_metadata_value attributes/proxy-uri | sed -e 's/^https:\/\///')
+readonly GRPC_LISTENING_PORT=$(/usr/share/google/get_metadata_value attributes/grpc-listening-port || echo "9085")
+# Point Dataproc components at this port.
+readonly THRIFT_LISTENING_PORT=$(/usr/share/google/get_metadata_value attributes/thrift-listening-port || echo "9083")
```
Comment (medb) [inline metastore-grpc-proxy/metastore-grpc-proxy.sh:29]: Too generic names

File: metastore-grpc-proxy/metastore-grpc-proxy.sh [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+# This is what the G2T proxy connects to. Set this to the DPMS Thrift endpoint.
+readonly HIVE_METASTORE_URI=$(/usr/share/google/get_metadata_value attributes/hive-metastore-uri || echo "$(hostname):9083")
+readonly HIVE_VERSION=$(/usr/share/google/get_metadata_value attributes/hive-version)
+readonly DOCKER_IMAGE=$(/usr/share/google/get_metadata_value attributes/docker-image || echo "gcr.io/cloud-metastore-public/hms-grpc-proxy/${HIVE_VERSION}")
```
Comment (medb) [inline metastore-grpc-proxy/metastore-grpc-proxy.sh:33]: Too generic name

**Review by medb — CHANGES_REQUESTED** [review]

File: metastore-grpc-proxy/metastore-grpc-proxy.sh (hunk 2 above)
Comment (medb) [reply metastore-grpc-proxy/metastore-grpc-proxy.sh]: Can we still use `MetastoreConfig` but override necessary properties to work with gRPC proxy?

**Comment by medb** [comment]
Closing as obsolete

**Pull request closed without merging**